File: python/Brunel/_weightpath.py

__all__ = ["WeightPath"]


def _setState(state, val, default=None):
    try:
        result = state[val]
        if result:
            return result
        else:
            return default
    except Exception:
        return default


class WeightPath:
    """Holds information about the change of weight with respect to dates"""
    def __init__(self, props=None, getHook=None):
        self._getHook = getHook

        self.state = {
            "id": None,
            "start_weight": None,
            "start_date": None,
            "weight_change": None,
            "date_change": None,
            "location": None,
        }

        self.setState(props)

    def __str__(self):
        return f"Source({self.getName()})"

    def __repr__(self):
        return self.__str__()

    def getID(self):
        return self.state["id"]

    def getStartDate(self):
        return self.state["start_date"]

    def getDateChange(self):
        return self.state["date_change"]

    def getStartWeight(self):
        return self.state["start_weight"]

    def getWeightChange(self):
        return self.state["weight_change"]

    def getLocation(self):
        return self.state["location"]

    def setState(self, state):
        if not state:
            return

        from ._daterange import Date as _Date

        self.state["id"] = _setState(state, "id")
        self.state["start_date"] = _setState(state, "start_date", None)
        self.state["start_weight"] = _setState(state, "start_weight", None)
        self.state["date_change"] = _setState(state, "date_change", None)
        self.state["weight_change"] = _setState(state, "weight_change", None)
        self.state["location"] = _setState(state, "location", None)

    def toDry(self):
        return self.state

    @staticmethod
    def unDry(value):
        return WeightPath(value)

    @staticmethod
    def load(data):
        return WeightPath(data)

File: src/model/WeightPath.js
import Dry from "json-dry";
import lodash from "lodash";
import KeyDate from "./KeyDate";

function setState(val, def = null) {
  if (val) {
    return val;
  } else {
    return def;
  }
}

class WeightPath {
  constructor(props) {
    this.state = {
      id: null,
      start_weight: null,
      start_date: null,
      weight_change: null,
      date_change: null,
      location: null,
    };

    this.setState(props);

    this._getHook = null;
    this._isAWeightPathObject = true;
  }

  static clone(item) {
    let c = new WeightPath();
    c._getHook = item._getHook;
    c.state = lodash.cloneDeep(item.state);
    return c;
  }

  getID() {
    return this.state.id;
  }

  getStartWeight() {
    if (!this.state.start_weight){
      return 0.0;
    } else {
      return this.state.start_weight;
    }
  }

  getStartDate() {
    return this.state.start_date;
  }

  getWeightChange() {
    if (!this.state.weight_change){
      return 0.0;
    } else {
      return this.state.weight_change;
    }
  }

  getDateChange() {
    return this.state.date_change;
  }

  getLocation() {
    return this.state.location;
  }

  getWeightAtDate(keydate){
    if (!keydate){
      return this.getStartWeight();
    }

    let d = this.getStartDate();

    if (d === null){
      return this.getStartWeight();
    }

    d = new KeyDate({name: d});

    if (d.isLater(keydate)){
      return 0.0;
    } 
    
    d = this.getDateChange();

    if (d === null){
      return this.getStartWeight();
    }

    d = new KeyDate({name: d});

    if (d.isLater(keydate)) {
      return this.getStartWeight();
    } else { 
      return this.getStartWeight() + this.getWeightChange();
    }
  }

  setState(state) {
    if (state) {
      this.state.id = setState(state.id);
      this.state.start_weight = setState(parseFloat(state.start_weight), 1.0);
      this.state.start_date = setState(parseFloat(state.start_date), null);
      this.state.weight_change = setState(parseFloat(state.weight_change), 0.0);
      this.state.date_change = setState(parseFloat(state.date_change), null);
      this.state.location = setState(state.location, null);
    }
  }

  _updateHooks(hook) {
    this._getHook = hook;
  }

  merge() {
    return this;
  }

  toString() {
    return `WeightPath(${this.getStartWeight()}-${this.getStartDate()})`;
  }

  toDry() {
    return { value: this.state };
  }
}

WeightPath.unDry = function (value) {
  return new WeightPath(value);
};

Dry.registerClass("WeightPath", WeightPath);

export default WeightPath;
